src/gpio.rs

```diff
@@ -293,1040 +377,841 @@ macro_rules! gpio {
+        unsafe { (*Gpio::<P>::ptr()).bsrr.write(|w| w.bits(1 << (16 + N))) }
+    }
 
-                    /// Configures the pin to operate as an open-drain output pin.
-                    /// Initial state will be low.
-                    #[inline]
-                    pub fn into_open_drain_output(
-                        self,
-                        cr: &mut $CR,
-                    ) -> $PXi<Output<OpenDrain>> {
-                        self.into_open_drain_output_with_state(cr, State::Low)
-                    }
+    #[inline(always)]
+    fn _is_set_low(&self) -> bool {
+        unsafe { (*Gpio::<P>::ptr()).odr.read().bits() & (1 << N) == 0 }
+    }
+    #[inline(always)]
+    fn _is_low(&self) -> bool {
+        // NOTE(unsafe) atomic read with no side effects
+        unsafe { (*Gpio::<P>::ptr()).idr.read().bits() & (1 << N) == 0 }
+    }
+}
 
-                    /// Configures the pin to operate as an open-drain output pin.
-                    /// `initial_state` specifies whether the pin should be initially high or low.
-                    #[inline]
-                    pub fn into_open_drain_output_with_state(
-                        mut self,
-                        cr: &mut $CR,
-                        initial_state: State,
-                    ) -> $PXi<Output<OpenDrain>> {
-                        self.set_state(initial_state);
-                        unsafe {
-                            $PXi::<Output<OpenDrain>>::set_mode(cr)
-                        }
-                    }
-                    /// Configures the pin to operate as an push-pull output pin.
-                    /// Initial state will be low.
-                    #[inline]
-                    pub fn into_push_pull_output(
-                        self,
-                        cr: &mut $CR,
-                    ) -> $PXi<Output<PushPull>> {
-                        self.into_push_pull_output_with_state(cr, State::Low)
-                    }
+impl<MODE, CR, const P: char, const N: u8> Pin<MODE, CR, P, N>
+where
+    MODE: Active,
+{
+    /// Erases the pin number from the type
+    #[inline]
+    pub fn erase_number(self) -> PEPin<MODE, P> {
+        PEPin::new(N)
+    }
+}
 
-                    /// Configures the pin to operate as an push-pull output pin.
-                    /// `initial_state` specifies whether the pin should be initially high or low.
-                    #[inline]
-                    pub fn into_push_pull_output_with_state(
-                        mut self,
-                        cr: &mut $CR,
-                        initial_state: State,
-                    ) -> $PXi<Output<PushPull>> {
-                        self.set_state(initial_state);
-                        unsafe {
-                            $PXi::<Output<PushPull>>::set_mode(cr)
-                        }
-                    }
+// embedded_hal impls
+impl<MODE, CR, const P: char, const N: u8> Pin<Output<MODE>, CR, P, N> {
```

Comment:
I haven't tested this, but what happens if you `use embedded_hal::digital::v2::OutputPin`, will the compiler not throw an "ambiguous function call error" since we have both the local `set_high` and the OutputPin set_high in scope.

Is this a problem? Not having to `unwrap` certainly has its advantages, but at the same time, it would make code like

```rust
use embedded_hal::digital::v2::OutputPin;

fn generic_fn(pin: impl OutputPin) {
    // ...
}

fn main() {
    let pin = // get a pin
    pin.set_high(); // Now ambiguous?
}
```

break, which might be confusing if the generic stuff is added later in a project.

Ok, I tested it in the playground and it seems like this is a non-issue. I guess I'm not entirely sure how method resolution works in rust.. I'll leave this comment here in case it is an issue in this case, but feel free to ignore it.